### PR TITLE
Synchronous write to tag views table

### PR DESCRIPTION
**Pull request by chbatey**
#259 

This changes the tag write to be part of the Future returned from asyncWriteMessages

It is still asynchronous with respect to the write to the messages table so it can still fail/timeout while the messages table write succeed however now the actor will be stopped. The tag_view will then be fixed when it is restarted.

Note that the TagWriter does keep retrying the write until the persistent actor stops, need to decide what to set that timeout to.

### CODE REVIEW

**Comment by chbatey** [comment]
Found some issues during load testing so hold off reviewing this one for now

**Review by patriknw — COMMENTED** [review]
Looked briefly at this and seems like a good approach. I will try to review more thoroughly next week. It's a substantial change but still not as scary as I expected.

File: core/src/main/scala/akka/persistence/cassandra/journal/TagWriter.scala
Comment (patriknw) [inline core/src/main/scala/akka/persistence/cassandra/journal/TagWriter.scala:369]: still TODO?

File: core/src/main/scala/akka/persistence/cassandra/journal/TagWriter.scala
Comment (patriknw) [inline core/src/main/scala/akka/persistence/cassandra/journal/TagWriter.scala:376]: is this logged for each write? too much? (do we have a verbose logging flag here?)

File: core/src/main/scala/akka/persistence/cassandra/journal/TagWriter.scala
Comment (patriknw) [inline core/src/main/scala/akka/persistence/cassandra/journal/TagWriter.scala:389]: does the buffer still contain the `events` or is this that we don't retry them in this way any more?

File: core/src/main/scala/akka/persistence/cassandra/journal/TagWriter.scala
Comment (patriknw) [inline core/src/main/scala/akka/persistence/cassandra/journal/TagWriter.scala:305]: nice

File: core/src/main/scala/akka/persistence/cassandra/journal/TagWriter.scala
Comment (patriknw) [inline core/src/main/scala/akka/persistence/cassandra/journal/TagWriter.scala:466]: ^ ?

File: core/src/main/scala/akka/persistence/cassandra/journal/TagWriters.scala
Comment (patriknw) [inline core/src/main/scala/akka/persistence/cassandra/journal/TagWriters.scala:225]: `pipeTo(sender())` would be ok

File: core/src/main/scala/akka/persistence/cassandra/journal/TagWriters.scala
Comment (patriknw) [inline core/src/main/scala/akka/persistence/cassandra/journal/TagWriters.scala:228]: same here

File: core/src/main/scala/akka/persistence/cassandra/journal/TagWriters.scala
Comment (patriknw) [inline core/src/main/scala/akka/persistence/cassandra/journal/TagWriters.scala:339]: do we still need `tellTagActor` or can be simplify by changing to askTagActor for the remaining usage of tellTagActor?

File: core/src/test/scala/akka/persistence/cassandra/journal/TagWritersSpec.scala
Comment (patriknw) [inline core/src/test/scala/akka/persistence/cassandra/journal/TagWritersSpec.scala:22]: ^ ?

**Review by patriknw — COMMENTED** [review]
looking very good

File: core/src/main/scala/akka/persistence/cassandra/journal/TagWriter.scala
Comment (chbatey) [reply core/src/main/scala/akka/persistence/cassandra/journal/TagWriter.scala:114]: 👍

File: core/src/main/scala/akka/persistence/cassandra/journal/TagWriter.scala
Comment (chbatey) [reply core/src/main/scala/akka/persistence/cassandra/journal/TagWriter.scala:335]: Ya definitely. Removing this was just from when i was debugging

File: core/src/main/scala/akka/persistence/cassandra/journal/TagWriter.scala
Comment (chbatey) [reply core/src/main/scala/akka/persistence/cassandra/journal/TagWriter.scala:312]: Done

**Comment by chbatey** [comment]
Also refs https://github.com/akka/akka-persistence-cassandra/issues/405

File: core/src/main/scala/akka/persistence/cassandra/journal/CassandraJournal.scala
Comment (chbatey) [reply core/src/main/scala/akka/persistence/cassandra/journal/CassandraJournal.scala:287]: Doing now

File: core/src/main/scala/akka/persistence/cassandra/journal/TagWriter.scala
Comment (chbatey) [reply core/src/main/scala/akka/persistence/cassandra/journal/TagWriter.scala:369]: The places we didn't batch correctly are now both fixed as part of this PR

File: core/src/main/scala/akka/persistence/cassandra/journal/TagWriter.scala
Comment (chbatey) [reply core/src/main/scala/akka/persistence/cassandra/journal/TagWriter.scala:389]: The buffer still contains the events, they are only removed once a writeComplete happens

File: core/src/main/scala/akka/persistence/cassandra/journal/TagWriters.scala
Comment (chbatey) [reply core/src/main/scala/akka/persistence/cassandra/journal/TagWriters.scala:339]: Nope! Seems we don't ever send a TagFlush to TagWriters anyhow

File: core/src/main/resources/reference.conf
Comment (chbatey) [reply core/src/main/resources/reference.conf]: I've added the explanation to the new timeout setting

File: core/src/main/scala/akka/persistence/cassandra/journal/TagWriter.scala
Comment (chbatey) [reply core/src/main/scala/akka/persistence/cassandra/journal/TagWriter.scala:254]: Removed, was when debugging

File: core/src/main/scala/akka/persistence/cassandra/journal/TagWriters.scala
Comment (johanandren) [inline core/src/main/scala/akka/persistence/cassandra/journal/TagWriters.scala:348]: Another map that can be on parasitic ec

File: core/src/main/scala/akka/persistence/cassandra/journal/TagWriters.scala
Comment (johanandren) [inline core/src/main/scala/akka/persistence/cassandra/journal/TagWriters.scala:340]: This is called from a future callback but uses the actor logger and may call updatePendingScanning which mutates the actor state.

Comment (chbatey) [reply core/src/main/scala/akka/persistence/cassandra/journal/TagWriters.scala:340]: Good catch!

File: core/src/main/scala/akka/persistence/cassandra/journal/CassandraJournal.scala
Comment (chbatey) [reply core/src/main/scala/akka/persistence/cassandra/journal/CassandraJournal.scala:285]: i'll move it inside

File: core/src/main/scala/akka/persistence/cassandra/journal/TagWriters.scala
Comment (chbatey) [reply core/src/main/scala/akka/persistence/cassandra/journal/TagWriters.scala:340]: Actually, the traverse calls this on the calling thread but i think doing that explicitly is more clear

**Comment by chbatey** [comment]
@patriknw do you mind taking one last look at this before we merge?

**Review by patriknw — APPROVED** [review]
LGTM (I looked at the additional review commits)

**Review by johanandren — APPROVED** [review]

File: core/src/main/resources/reference.conf
Comment (agapic) [inline core/src/main/resources/reference.conf:255]: Just came across this PR when debugging an issue: https://discuss.lightbend.com/t/akka-persistence-cassandra-unrecoverable-failures-when-tag-writes-buffer-gets-too-big/7643

We're using version 1.0.1 of this, would upgrading to 1.0.4 help resolve the issue we're seeing?

Comment (chbatey) [reply core/src/main/resources/reference.conf:255]: Replied on the thread

**Pull request merged**
